src/deps-index.ts

```diff
@@ -21,79 +21,100 @@ export interface DepsIndex {
-      const lazyImports = new Set<UrlString>();
-      for (const htmlImport of imports) {
-        try {
-          console.assert(
-              htmlImport.url, 'htmlImport: %s has no url', htmlImport);
-        } catch (err) {
-          if (err instanceof AssertionError) {
-            continue;
-          }
-          throw err;
-        }
+/**
+ * For a given URL, return a set of eager and lazy dependencies.  The
+ * function is recursive, but a `visited` set is carried forward to short-
+ * circuit loops encountered.
+ */
+function getTransitiveDependencies(
+    url: UrlString,
+    analysis: Analysis,
+    visited: Set<UrlString> = new Set<UrlString>()): DependencyMapEntry {
+  const map: DependencyMapEntry = {
+    eager: new Set(),
+    lazy: new Set(),
+    lazyEdges: new Set()
+  };
+  const document = analysis.getDocument(url);
+  if (!(document instanceof Document)) {
+    const message = document && document.message || 'unknown';
+    throw new Error(`Unable to get document ${url}: ${message}`);
+  }
 
-        switch (htmlImport.type) {
-          case 'html-import':
-            eagerImports.add(htmlImport.document.url);
-            break;
-          case 'lazy-html-import':
-            lazyImports.add(htmlImport.document.url);
-            break;
-        }
-      }
-      return {url: url, eager: eagerImports, lazy: lazyImports};
+  // Gets a collection of only the direct html imports for the document.
+  const htmlImports = document.getFeatures(
+      {kind: 'html-import', imported: false, externalPackages: true});
+
+  // Initially, we will place each import into the dependency map based on
+  // whether it was explicitly a lazy import.
+  htmlImports.forEach((i) => {
+    if (i.lazy) {
+      map.lazy.add(i.document.url);
+      map.lazyEdges.add(i.document.url);
+    } else {
+      map.eager.add(i.document.url);
     }
+  });
+
+  visited.add(url);
+
+  // Every unvisited direct eager dependency's lazy dependencies are recognized
+  // as transitively lazy unless already recognized as eager.  All others are
+  // treated as eager.
+  Array.from(map.eager).filter((d) => !visited.has(d)).forEach((d) => {
+    const deps = getTransitiveDependencies(d, analysis, visited);
+    deps.eager.forEach((dep) => map.eager.add(dep) && map.lazy.delete(dep));
+    deps.lazy.forEach((dep) => map.eager.has(dep) || map.lazy.add(dep));
+    deps.lazyEdges.forEach((dep) => map.lazyEdges.add(dep));
+  });
+
+  // Every unvisited direct lazy dependency's dependencies are recognized
+  // as transitively lazy unless already defined as eager.
+  Array.from(map.lazy).filter((d) => !visited.has(d)).forEach((d) => {
+    const deps = getTransitiveDependencies(d, analysis, visited);
+    deps.eager.forEach(
+        (dep) => map.eager.has(dep) ? map.lazy.delete(dep) : map.lazy.add(dep));
+    deps.lazy.forEach((dep) => map.eager.has(dep) || map.lazy.add(dep));
+    deps.lazyEdges.forEach((dep) => map.lazyEdges.add(dep));
+  });
```

Comment:
Some light afternoon reading: https://www.kent.edu/sites/default/files/TR-KSU-CS-2010-01.pdf

We should be able to do this in a single BFS, with a change that a node can be visited more than once if we're seeing it via an eager edge for the first time, essentially performing two reachability queries at the same time:

Something like this modified BFS should work:

```js
getHtmlDependencies(document) {
  const deps = new Set();
  const eagerDeps = new Set();
  const lazyImports = new Set();
  _getHtmlDependencies(document, true, deps, eagerDeps, lazyImports);
  return {deps, eagerDeps, lazyImports};
}

_getHtmlDependencies(document, viaEager, visited, visitedEager, lazyImports) {
  const imports = document.getFeatures({kind: 'html-import', imported: false, externalPackages: true});
  for (const import of imports) {
    const importUrl = import.document.url;
    if (import.lazy) {
      lazyImports.add(importUrl);
    }
    if (visitedEager.has(importUrl)) {
      continue;
    }
    const isEager = viaEager || !import.lazy;
    if (isEager) {
      visitedEager.add(importUrl);
      // In this case we've visited a node eagerly for the first time,
      // so recurse
    } else if (visited.has(importUrl)) {
      // In this case we're seeing a node lazily again, so don't recurse
      continue;
    }
    visited.add(importUrl);
    getHtmlDependencies(import.document, isEager, visited, visitedEager, lazyImports);
  }
}
```